planner/core/plan_cache_lru.go

```diff
@@ -0,0 +1,190 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package core
+
+import (
+	"container/list"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/util/kvcache"
+)
+
+// CacheEntry wraps Key and Value. It's the value of list.Element.
+type CacheEntry struct {
+	PlanKey   kvcache.Key
+	PlanValue kvcache.Value
+}
+
+// LRUPlanCache is a simple least recently used cache, not thread-safe, JUST use for plan cache.
+type LRUPlanCache struct {
+	capacity uint
+	size     uint
+	// buckets replace the map in general LRU
+	buckets map[string][]*list.Element
+
+	// pickFromBucket set rules to get one element from bucket.
+	pickFromBucket func([]*list.Element, interface{}) (*list.Element, int, bool)
+
+	// onEvict function will be called if any eviction happened
+	onEvict func(kvcache.Key, kvcache.Value)
+	cache   *list.List
+}
+
+// NewLRUPlanCache creates a PCLRUCache object, whose capacity is "capacity".
+// NOTE: "capacity" should be a positive value.
+func NewLRUPlanCache(capacity uint) *LRUPlanCache {
+	if capacity < 1 {
+		panic("capacity of LRU Cache should be at least 1.")
+	}
+	return &LRUPlanCache{
+		capacity: capacity,
+		size:     0,
+		buckets:  make(map[string][]*list.Element),
+		onEvict:  nil,
+		cache:    list.New(),
+	}
+}
+
+// SetOnEvict set the function called on each eviction.
+func (l *LRUPlanCache) SetOnEvict(onEvict func(kvcache.Key, kvcache.Value)) {
+	l.onEvict = onEvict
+}
+
+// SetPickFromBucket set the function called on each eviction.
+func (l *LRUPlanCache) SetPickFromBucket(pick func([]*list.Element, interface{}) (*list.Element, int, bool)) {
+	l.pickFromBucket = pick
+}
+
+// Get tries to find the corresponding value according to the given key.
+func (l *LRUPlanCache) Get(key kvcache.Key, paramTypes interface{}) (value kvcache.Value, ok bool) {
+	bucket, bucketExist := l.buckets[string(key.Hash())]
+	if bucketExist {
+		if element, _, exist := l.pickFromBucket(bucket, paramTypes); exist {
+			l.cache.MoveToFront(element)
+			return element.Value.(*CacheEntry).PlanValue, true
+		}
+	}
+	return nil, false
+}
+
+// Put puts the (key, value) pair into the LRU Cache.
+func (l *LRUPlanCache) Put(key kvcache.Key, value kvcache.Value, paramTypes interface{}) {
+	hash := string(key.Hash())
+	bucket, bucketExist := l.buckets[hash]
+	// bucket exist
+	if bucketExist {
+		if element, _, exist := l.pickFromBucket(bucket, paramTypes); exist {
+			element.Value.(*CacheEntry).PlanValue = value
+			l.cache.MoveToFront(element)
+			return
+		}
```

Comment:
What if the bucket exists but we cannot pick a valid element from the current bucket?